jack-core/src/com/rapleaf/jack/transaction/DbMetricsImpl.java

```diff
@@ -0,0 +1,157 @@
+    this.maxTotalConnections = maxTotalConnections;
+    this.minIdleConnections = minIdleConnections;
+    this.maxWaitMillis = maxWaitMillis;
+    this.keepAliveMillis = keepAliveMillis;
+  }
+
+  void update(boolean isOpenConnection, final GenericObjectPool connectionPool) {
+    try {
+      long updateToNowTime = lifeTimeStopwatch.elapsedMillis() - lastUpdateTime;
+      lastUpdateTime = lifeTimeStopwatch.elapsedMillis();
+
+      int numActive = connectionPool.getNumActive();
+      int numIdle = connectionPool.getNumIdle();
+      int numWaiters = connectionPool.getNumWaiters();
+
+      maxConnectionWaitTime = connectionPool.getMaxBorrowWaitTimeMillis();
+      totalActiveTime += numActive * updateToNowTime;
+      totalIdleTime += numIdle * updateToNowTime;
+      totalConnectionWaitTime += numWaiters * updateToNowTime;
+
+      if (isOpenConnection) {
+        totalConnections += 1;
+      }
+      int newConnectionsNumber = numActive + numIdle;
+      if (newConnectionsNumber > currentConnectionsNumber) {
+        openedConnectionsNumber += newConnectionsNumber - currentConnectionsNumber;
+      }
+      currentConnectionsNumber = newConnectionsNumber;
+
+
+      if (numActive == maxTotalConnections) {
+        maxActiveConnectionsTime += updateToNowTime;
+      }
+
+      if ((numIdle + numActive == minIdleConnections) && (numIdle > 0)) {
+        minIdleConnectionsTime += updateToNowTime;
+      }
+    } catch (Exception e) {
+      LOG.error("failed to update statistics", e);
+    }
+  }
+
+  @Override
+  public long getTotalConnections() {
+    return totalConnections;
+  }
+
+  @Override
+  public double getMaxConnectionsProportion() {
+    return (double)maxActiveConnectionsTime / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public double getAverageConnectionWaitTime() {
+    return (double)totalConnectionWaitTime / (double)totalConnections;
+  }
+
+  @Override
+  public double getAverageConnectionExecutionTime() {
+    return ((double)totalActiveTime / (double)totalConnections);
+  }
+
+  @Override
+  public double getMinConnectionsProportion() {
+    return (double)minIdleConnectionsTime / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public long getOpenedConnectionsNumber() {
+    return openedConnectionsNumber;
+  }
+
+  @Override
+  public double getAverageIdleConnections() {
+    return (double)totalIdleTime / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public double getAverageActiveConnections() {
+    return (double)totalActiveTime / (double)lifeTimeStopwatch.elapsedMillis();
+  }
+
+  @Override
+  public long getMaxConnectionWaitingTime() {
+    return maxConnectionWaitTime;
+  }
+
+  @Override
+  public String log() {
+    String log = "";
+
+    log += ("\n-----------------------TRANSACTOR METRICS-----------------------\n");
+    log += ("\nAverage (over time) number of Idle connections : " + String.valueOf(getAverageIdleConnections()));
```

Comment:
For double numbers, can you format them so that they only show two digits? Can use the `String.format` method.